### ISSUE TITLE
Consistency in collections API for partitions

### ISSUE THREAD

**Issue by koverholt** [dask/dask#1008]
I have a few points of feedback after working across different dask collections for benchmarks and examples.

Currently, dask collections have a partitioning argument (depending on the particular method being used) and an attribute to show the number of partitions:

| Collection | Specify partitions | Query number of partitions |
| --- | --- | --- |
| dask.array | chunks | numblocks |
| dask.bag | npartitions/chunkbytes/partition_size | npartitions |
| dask.dataframe | npartitions/chunksize | npartitions |

1) When moving between different `from_*` methods in arrays and bags/dataframes, I often pause to remember which argument is used to specify the number of partitions or chunk sizes. Thoughts on making the number of partitions a consistent argument (ideally, `partitions=X` or `npartitions=X`) between all collections? For arrays, I almost always find myself inverting for the chunk sizes to arrive at a particular number of partitions (especially when I'm using the distributed scheduler). It would be comforting to know that I can create a collection and always use a `partitions` or `npartitions` argument whether I'm using `dask.bag.from_filenames` or `dask.array`, any reason this might not be feasible for all collections?

2) Assuming #1 is a valid point, thoughts on making the argument to specify partitions the same as the attribute for querying the number of partitions? For `dask.array`, I need to remember `chunks` vs. `numblocks`.

3) It might be handy to show the number of partitions in the repr for each collection, e.g.:

```
>>> x
dask.array<arange-..., shape=(100,), dtype=int64, chunksize=(2,), partitions=10>

>>> y
dd.Series<from-dask-arraycdefdd873cdf06131bcc8da67a0e65fb, divisions=(0, 2, 4, ..., 98, 99), partitions=10>

>>> b
<dask.bag.core.Bag at 0x11312e0f0, partitions=10>
```

**Comment by mrocklin** [comment]
Unifying the interface sounds good.  We can do this at the superclass level to ensure consistency

Note however that arrays are different, there is more structure here than just the number of partitions, there is how the data is partitioned across many dimensions.  Numblocks/chunks will likely have to stay around; they mean something somewhat different than npartitions.

**Comment by koverholt** [comment]
Thanks for the feedback. I wasn't suggesting doing away with numblocks/chunks for arrays for the reason you mentioned, but rather a superclass would be a handy way to work across collections.

**Comment by mrocklin** [comment]
OK so there are a few goals here:

1.  Use `npartitions=` in most data ingest functions:  This turns out to be fairly difficult to do universally.  Almost all data ingest systems think about the size that they're going to break things up by and each system has a different way of determining the size of the whole thing.  This is definitely doable, but it's not clear that the work/payoff ratio makes it worth it.
2.  Report `npartitions` more widely:  I've added this to the dataframe repr (definitely more informative that divisions in the common case) but adding it to the dask.array repr pushes us past 80 characters.  Also, I think that almost all dask.array users are more comfortable with chunks than with npartitions.  Chunks has been around in that community for a long time and is surprisingly natural on the ground.
3.  Have an npartitions attribute everywhere.  Done.

For arrays generally my suggestion is not to care.  I don't think that consistency between array and bag/dataframe is particularly useful here.  I think that it's more important for dask.array to align with other array tools and for bag/dataframe to align with other bag/dataframe tools.  I suspect that you're in the artificial position of having to deal with everything without being married to a particular application.

So I'm going to punt on full adherence here.  Are there things that are particularly glaring that can be easily fixed.  I've done a bit of work at https://github.com/dask/dask/pull/1142

**Comment by koverholt** [comment]
Thanks for the reasoning around data ingest and chunks.

+1 on the changes in #1142, I gave it a spin. Having the `npartitions` attribute handy across collections is nice, and the `repr` for dataframe showing `npartitions` is also nice. Thanks!

**Comment by jcrist** [comment]
I think this was resolved in #1142. Closing for now, feel free to reopen if there's still more to do.
